tendenci/apps/payments/paypal/utils.py

```diff
@@ -119,8 +120,28 @@ def verify_no_fraud(response_d, payment):
     txn_id = response_d.get('txn_id')
     if not txn_id:
         return False
-
-    txn_id_exists = Payment.objects.filter(trans_id=txn_id).exists()
+    # To prevent a single txn_id from being used for multiple Payments by
+    # simultaneously submitting multiple requests using the same txn_id, the
+    # database record for the current Payment object must be updated with the
+    # txn_id before checking for duplicates, and the database transaction
+    # associated with this update must be committed without being nested in any
+    # other transactions.
+    # transaction.commit() ensures that this transaction is not nested in any
+    # other transactions by committing the current transaction (if autocommit is
+    # disabled), by throwing an exception (if called within a transaction.atomic
+    # block), or by doing nothing (if this transaction is not nested).
+    # transaction.atomic() with select_for_update() ensures that
+    # payment.trans_id cannot be updated in the database by another process
+    # after this calls get_object_or_404() and before this calls payment.save().
+    transaction.commit()
+    with transaction.atomic():
+        payment = get_object_or_404(Payment.objects.select_for_update(), pk=payment.id)
+        if payment.trans_id != '':
+            if payment.trans_id != txn_id:
+                return False
+            payment.trans_id = txn_id
```

Comment:
At this point, it's unsure if this txn_id is legitimate. Should it really be saved? What if a valid later transaction has this txn_id?

Reply:
Good point.  This code was designed to ensure that invalid requests are never accepted, and it was not designed to ensure that valid requests are always accepted.  However, we could move the duplicate txn_id check to the end so that it is the last check before writing the database record.